Q1 added

matinehAkhlaghinia opened the pull request:
An algorithm which decides if one string is a permutation of the other is implemented.

matinehAkhlaghinia committed:
Add files via upload

Brusi commented on Q1.cpp:
```diff
@@ -0,0 +1,52 @@
+using namespace std;
+
+
+bool mapIsEmpty(unordered_map<char, int> freq_map) {
```
The map can be heavy, better to pass as const reference: 

`bool mapIsEmpty(const unordered_map<char, int>& freq_map) { .. }`

Brusi replied:
`mapIsEmpty` is not such a good name, since the map is not empty - it has values that are zeros. Can be confused with calling `freq_map.empty()` which is different.
I suggest calling it 'mapHasOnlyZeros' or anything that states exactly what it does.

Brusi commented on Q1.cpp:
```diff
@@ -0,0 +1,52 @@
+  return true;
+}
+
+bool isPermutation(string s1, string s2) {
```
If you don't change the strings, consider passing them as const reference to avoid copying the strings:
bool validInput(const string& s1, const string& s2){ .. }

Brusi commented on Q1.cpp:
```diff
@@ -0,0 +1,52 @@
+    frequency_map[s1[i]]++;
+  }
+  for(int j = 0; j < s2.length(); j++) {
+  	if(frequency_map.find(s2[j]) != frequency_map.end())
```
I see you made this optimization: if a letter of the second word not exist in the first one, return false immediately.
Of course it was not mandatory, since the `mapIsEmpty` check at the end would catch negative numbers too, but it is nice to have.

I suggest a similar optimization (that will catch more cases):
```
for (int j=0; j < s2.length(); j++) {
    frequency_map[s2[j]]--;
    if (frequency_map[s2[j]] < 0) return false;
}
```
Like this you will catch more cases and have the same number of accesses to the map (2).

But of course, your optimization is good too, and neither is mandatory :)

Brusi commented on Q1.cpp:
```diff
@@ -0,0 +1,52 @@
+
+bool isPermutation(string s1, string s2) {
+  unordered_map<char, int> frequency_map;
+  transform(s1.begin(), s1.end(), s1.begin(), ::tolower); 
```
Nice use of std::transform :+1:

Brusi commented on Q1.cpp:
```diff
@@ -0,0 +1,52 @@
+  return mapIsEmpty(frequency_map);
+}
+
+int main() {
```
:+1: Very short code! Good work!

Brusi reviewed (approved):
Good work! Nice and short code :)
See comments.

matinehAkhlaghinia committed:
update#1

changes made according to the comments, passing arguments as const refs
and making the name of the method more appropriate

Brusi commented on Q1.cpp:
```diff
@@ -12,41 +12,41 @@ Apple, Pabble ⇒ False */
-  unordered_map<char, int>::iterator it;
+bool mapHasOnlyZeroes(const unordered_map<char, int>& freq_map) {
+  unordered_map<char, int>::const_iterator it;
   for(it = freq_map.begin(); it != freq_map.end(); it++) {
```
If you want, you can inline the iterator decleration into the `for` loop:
`for (unordered_map<char, int>::const_iterator it = freq_map.begin(); it != freq_map.end(); it++) { ... }`
The advantage of it is that the scope of `it` is inside the loop (cannot and should be used after the loop).

FYI, If C++11, you can use `auto` instead of the long, tedious iterator type `unordered_map<char, int>::iterator`:
`for (auto it = freq_map.begin(); it != freq_map.end(); it++) { ... }`
or even use 'foreach' loops:
```
for (const auto& entry : freq_map) {
    if (entry.second != 0) { ...
    ...
}
```

Brusi reviewed (approved):
Well done :)

(Fixing the code from this point is not mandatory, since Assignment 1 has only one review iteration. My comments here are for your general knowledge :) )
